### PR TITLE
congestion control

### PR DESCRIPTION
**Pull request by kazuho**

### CODE REVIEW

File: lib/quicly.c [hunk 1]
```diff
@@ -1118,6 +1162,9 @@ static quicly_conn_t *create_connection(quicly_context_t *ctx, const char *serve
     quicly_loss_init(&conn->egress.loss, conn->super.ctx->loss,
                      conn->super.ctx->loss->default_initial_rtt /* FIXME remember initial_rtt in session ticket */);
     conn->egress.path_challenge.tail_ref = &conn->egress.path_challenge.head;
+    conn->egress.send_ack_at = INT64_MAX;
+    update_cc_ticks(conn->super.ctx);
+    cc_init(&conn->egress.cc.ccv, &newreno_cc_algo, 1280 * 8, 1280);
```
Comment (kazuho) [inline lib/quicly.c]: The algorithm must be specified as part of the context rather than hard-coding.

Comment (janaiyengar) [inline lib/quicly.c]: Is newreno_cc_algo a function?

File: lib/quicly.c [hunk 2]
```diff
@@ -209,6 +211,24 @@ struct st_quicly_conn_t {
...
+         */
+        struct {
+            struct cc_var ccv;
+            size_t bytes_in_flight;
+            uint64_t end_of_recovery;
+            /**
+             * collected by on_ack_cc
+             */
+            struct {
+                size_t nsegs, nbytes;
```
Comment (janaiyengar) [inline lib/quicly.c]: Do these vars refer to the number of packets/segs and bytes newly acknowledged?

File: lib/quicly.c [hunk 3]
```diff
@@ -1622,6 +1677,8 @@ static int commit_send_packet(quicly_conn_t *conn, struct st_quicly_send_context
 {
     assert(s->target.cipher->aead != NULL);
 
+    assert(s->dst != s->dst_encrypt_from);
```
Comment (janaiyengar) [inline lib/quicly.c]: What does this assertion check for?

File: lib/quicly.c [hunk 4]
```diff
@@ -1752,6 +1821,10 @@ static int prepare_acked_packet(quicly_conn_t *conn, struct st_quicly_send_conte
 
     if ((ret = prepare_packet(conn, s, min_space)) != 0)
         return ret;
+    /* FIXME move the window size check into prepare_packet to avoid sending an empty QUIC packet */
+    /* TODO return the remaining window that the sender can use */
+    if (s->send_window < (s->dst - s->target.packet->data.base) + min_space)
```
Comment (janaiyengar) [inline lib/quicly.c]: What's min_space used for?

File: lib/quicly.c [hunk 5]
```diff
@@ -1947,19 +2020,18 @@ int retire_acks(quicly_conn_t *conn, size_t count, size_t epoch)
...
-                    skip = 0;
-                } else if (epoch == 1 &&
-                           (ack->acked == on_ack_stream && !stream_is_handshake_flow(ack->data.stream.stream_id, 3))) {
-                    /* NACK all application data upon 0-RTT denial */
-                    skip = 0;
-                } else {
-                    skip = 1;
-                }
+        int skip = 0;
+        if (epoch != SIZE_MAX) {
```
Comment (janaiyengar) [inline lib/quicly.c]: I don't think I follow this block below. What does this have to do with congestion control?

File: lib/quicly.c [hunk 6]
```diff
@@ -1573,6 +1620,13 @@ static int on_ack_stream_id_blocked(quicly_conn_t *conn, int acked, quicly_ack_t
     return 0;
 }
 
+static int on_ack_cc(quicly_conn_t *conn, int acked, quicly_ack_t *ack)
+{
+    conn->egress.cc.this_ack.nsegs += 1;
+    conn->egress.cc.this_ack.nbytes += ack->data.cc.bytes_in_flight;
```
Comment (janaiyengar) [inline lib/quicly.c]: If these variables are supposed to be what's newly acked by a received ack, I don't think this is correct.

File: lib/quicly.c [hunk 7]
```diff
@@ -1994,6 +2068,9 @@ static int do_detect_loss(quicly_loss_t *ld, int64_t now, uint64_t largest_acked
         quicly_acks_release(&conn->egress.acks, &iter);
         quicly_acks_next(&iter);
     }
+    conn->egress.cc.bytes_in_flight -= conn->egress.cc.this_ack.nbytes;
```
Comment (janaiyengar) [inline lib/quicly.c]: You should guard this. A massive bytes_in_flight will basically block the connection competely.
conn->egress.cc.bytes_in_flight -= (conn->egress.cc.bytes_in_flight < conn->egress.cc.this_ack.nbytes) ?
     conn->egress.cc.bytes_in_flight : conn->egress.cc.this_ack.nbytes;

File: lib/quicly.c [hunk 8]
```diff
@@ -1994,6 +2068,9 @@ static int do_detect_loss(quicly_loss_t *ld, int64_t now, uint64_t largest_acked
...
+    if (conn->egress.cc.this_ack.nbytes != 0 && conn->egress.loss.rto_count == 0)
+        cc_cong_signal(&conn->egress.cc.ccv, CC_FIRST_RTO, (uint32_t)conn->egress.cc.bytes_in_flight);
```
Comment (janaiyengar) [inline lib/quicly.c]: What does this function do?

File: lib/quicly.c [hunk 9]
```diff
@@ -2417,13 +2500,26 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
         packet_number += frame->gaps[gap_index];
     }
 
+    /* update bytes in pipe */
+    conn->egress.cc.bytes_in_flight -= conn->egress.cc.this_ack.nbytes;
```
Comment (janaiyengar) [inline lib/quicly.c]: guard this

File: lib/quicly.c [hunk 10]
```diff
@@ -2417,13 +2500,26 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
...
+
+    /* loss-detection, as well as verifying  */
     quicly_loss_on_ack_received(&conn->egress.loss, frame->largest_acknowledged,
                                 last_packet_sent_at <= now && packet_number >= frame->largest_acknowledged
                                     ? (uint32_t)(now - last_packet_sent_at)
                                     : UINT32_MAX);
+    int rto_verified = quicly_loss_on_packet_acked(&conn->egress.loss, frame->largest_acknowledged);
```
Comment (janaiyengar) [inline lib/quicly.c]: what is verified here?

File: lib/quicly.c (hunk 1 above)
Comment (kazuho) [reply lib/quicly.c]: It's a list of callback functions, defined at https://github.com/h2o/daemons-cc/blob/master/cc_newreno.c#L58-L64.

File: lib/quicly.c (hunk 2 above)
Comment (kazuho) [reply lib/quicly.c]: Yes.

File: lib/quicly.c (hunk 3 above)
Comment (kazuho) [reply lib/quicly.c]: It checks that we are not trying to commit a packet that does not have any payload. This should be guaranteed by `prepare_packet` when it allocates a buffer for a packet.

File: lib/quicly.c (hunk 4 above)
Comment (kazuho) [reply lib/quicly.c]: As discussed in https://github.com/h2o/quicly/pull/13/files/797eafb9bc4aecbc6ade78367c517e97d238af55#r196601841, we should not allocate a packet buffer when a frame cannot be written to that frame.

`min_space` is the minimum space that is required to write a frame.

File: lib/quicly.c (hunk 5 above)
Comment (kazuho) [reply lib/quicly.c]: We are (still) using a unified inflight packet map (i.e. `quicly_acks_t`).

This means that when certain epoch is being closed, we need to remove references to the packets sent in that epoch from the map. The other point is that when 0-RTT is rejected we want to pretend as if we received NACK for all data we sent using 0-RTT.

This function does that, when `epoch` is used (i.e. set to something other than `SIZE_MAX`).

File: lib/quicly.c (hunk 8 above)
Comment (kazuho) [reply lib/quicly.c]: My understanding is that it signals the congestion controller that we have observed first RTO.

File: lib/quicly.c (hunk 10 above)
Comment (kazuho) [reply lib/quicly.c]: It is a flag telling if `OnRetransmissionTimeoutVerified` needs to be invoked. Based on https://tools.ietf.org/html/draft-ietf-quic-recovery-12#section-3.5.6.

File: lib/quicly.c (hunk 5 above)
Comment (gatopeich) [reply lib/quicly.c]: I see SIZE_MAX used all around this project (and picotls) as a "None" value.
Why don't you just define a "NONE" constant for that? It would make these expressions easier to understand.

**Comment by kazuho** [comment]
This is still very rough but works. Let's merge and fix the issues.

**Comment by janaiyengar** [comment]
SGTM
